compiler/rustc_data_structures/src/sync/lock.rs

```diff
@@ -3,224 +3,257 @@
-        // state because it is a invariant of this type.
-        unsafe { self.lock.raw.unlock() };
+    /// The value representing a locked state for the `Cell`.
+    const LOCKED: bool = true;
+
+    /// A lock which only uses synchronization if `might_be_dyn_thread_safe` is true.
+    /// It implements `DynSend` and `DynSync` instead of the typical `Send` and `Sync`.
+    pub struct Lock<T> {
+        mode: LockMode,
+        data: UnsafeCell<T>,
     }
-}
 
-#[cfg(parallel_compiler)]
-union LockRawUnion {
-    /// Indicates if the cell is locked. Only used if `LockRaw.sync` is false.
-    cell: ManuallyDrop<Cell<bool>>,
+    impl<T> Lock<T> {
+        #[inline(always)]
+        pub fn new(inner: T) -> Self {
+            Lock {
+                mode: if unlikely(mode::might_be_dyn_thread_safe()) {
+                    // Create the lock with synchronization enabled using the `RawMutex` type.
+                    LockMode::Sync(RawMutex::INIT)
+                } else {
+                    // Create the lock with synchronization disabled.
+                    LockMode::NoSync(Cell::new(!LOCKED))
+                },
+                data: UnsafeCell::new(inner),
+            }
+        }
 
-    /// A lock implementation that's only used if `LockRaw.sync` is true.
-    lock: ManuallyDrop<parking_lot::RawMutex>,
-}
+        #[inline(always)]
+        pub fn into_inner(self) -> T {
+            self.data.into_inner()
+        }
 
-/// A raw lock which only uses synchronization if `might_be_dyn_thread_safe` is true.
-/// It contains no associated data and is used in the implementation of `Lock` which does have such data.
-///
-/// A manual implementation of a tagged union is used with the `sync` field and the `LockRawUnion` instead
-/// of using enums as it results in better code generation.
-#[cfg(parallel_compiler)]
-struct LockRaw {
-    /// Indicates if synchronization is used via `opt.lock` if true,
-    /// or if a non-thread safe cell is used via `opt.cell`. This is set on initialization and never changed.
-    sync: bool,
-    opt: LockRawUnion,
-}
+        #[inline(always)]
+        pub fn get_mut(&mut self) -> &mut T {
+            self.data.get_mut()
+        }
 
-#[cfg(parallel_compiler)]
-impl LockRaw {
-    fn new() -> Self {
-        if unlikely(super::mode::might_be_dyn_thread_safe()) {
-            // Create the lock with synchronization enabled using the `RawMutex` type.
-            LockRaw {
-                sync: true,
-                opt: LockRawUnion { lock: ManuallyDrop::new(parking_lot::RawMutex::INIT) },
+        /// This dispatches on the `LockMode` and gives access to its variants depending on
+        /// `assume`. If `no_sync` returns `None` this will panic.
+        ///
+        /// Safety
+        /// This method must only be called if `might_be_dyn_thread_safe` on lock creation matches
+        /// matches the `assume` argument.
+        #[inline(always)]
+        #[track_caller]
+        unsafe fn dispatch<R>(
```

Comment:
I would like using `if` statement in `lock` `lock_assume` of `Lock`, and `drop` of `LockGuard`. Although they use some similar code, it preserves the readability.